tarantool/ext_types/decimal.py

```diff
@@ -0,0 +1,226 @@
+#
+# |  4 bits           |  4 bits           |
+#    = the last digit    = nibble
+#
+# The final nibble represents the number’s sign:
+#
+#     0x0a, 0x0c, 0x0e, 0x0f stand for plus,
+#     0x0b and 0x0d stand for minus.
+
+EXT_ID = 1
+
+TARANTOOL_DECIMAL_MAX_DIGITS = 38
+
+def get_mp_sign(sign):
+    if sign == '+':
+        return 0x0c
+
+    if sign == '-':
+        return 0x0d
+
+    raise RuntimeError
+
+def add_mp_digit(digit, bytes_reverted, digit_count):
+    if digit_count % 2 == 0:
+        bytes_reverted[-1] = bytes_reverted[-1] | (digit << 4)
+    else:
+        bytes_reverted.append(digit)
+
+def check_valid_tarantool_decimal(str_repr, scale, first_digit_ind):
+# Decimal numbers have 38 digits of precision, that is, the total number of
+# digits before and after the decimal point can be 38. If there are more
+# digits arter the decimal point, the precision is lost. If there are more
+# digits before the decimal point, error is thrown.
+#
+# Tarantool 2.10.1-0-g482d91c66
+#
+# tarantool> decimal.new('10000000000000000000000000000000000000')
+# ---
+# - 10000000000000000000000000000000000000
+# ...
+#
+# tarantool> decimal.new('100000000000000000000000000000000000000')
+# ---
+# - error: '[string "return VERSION"]:1: variable ''VERSION'' is not declared'
+# ...
+#
+# tarantool> decimal.new('1.0000000000000000000000000000000000001')
+# ---
+# - 1.0000000000000000000000000000000000001
+# ...
+#
+# tarantool> decimal.new('1.00000000000000000000000000000000000001')
+# ---
+# - 1.0000000000000000000000000000000000000
+# ...
+#
+# In fact, there is also an exceptional case: if decimal starts with `0.`,
+# 38 digits after the decimal point are supported without the loss of precision. 
+#
+# tarantool> decimal.new('0.00000000000000000000000000000000000001')
+# ---
+# - 0.00000000000000000000000000000000000001
+# ...
+#
+# tarantool> decimal.new('0.000000000000000000000000000000000000001')
+# ---
+# - 0.00000000000000000000000000000000000000
+# ...
+    if scale > 0:
+        digit_count = len(str_repr) - 1 - first_digit_ind
+    else:
+        digit_count = len(str_repr) - first_digit_ind
+
+    if digit_count <= TARANTOOL_DECIMAL_MAX_DIGITS:
+        return True
+
+    if (digit_count - scale) > TARANTOOL_DECIMAL_MAX_DIGITS:
+        raise MsgpackError('Decimal cannot be encoded: Tarantool decimal ' + \
+                           'supports a maximum of 38 digits.')
+
+    starts_with_zero = str_repr[first_digit_ind] == '0'
+
+    if (    (digit_count > TARANTOOL_DECIMAL_MAX_DIGITS + 1) or \
+            (digit_count == TARANTOOL_DECIMAL_MAX_DIGITS + 1 \
+            and not starts_with_zero)):
+        warn('Decimal encoded with loss of precision: ' + \
+             'Tarantool decimal supports a maximum of 38 digits.',
+             MsgpackWarning)
+        return False
+
+    return True
+
+def strip_decimal_str(str_repr, scale, first_digit_ind):
+    assert scale > 0
+    # Strip extra bytes
```

Comment:
No dot at the end of the comment.

Reply:
Removed all dots since comments in the remaining code don't have one